core/WFCORE-3435_capability_requirements_deployments.adoc

```diff
@@ -0,0 +1,184 @@
+
+For services that are activated using MSC `ServiceActivator`, the API does not provide a way to get such a capability name.
+Deployments that are activating services must be able to work with capabilities too.
+
+== Issue Metadata
+
+* WildFly issue: https://issues.jboss.org/browse/WFCORE-3435
+* Affected projects or components:
+** https://github.com/wildfly/wildfly-core[Wildfly Core]
+** https://github.com/jboss-msc/jboss-msc[JBoss Modular Service Container]
+** https://github.com/wildfly-swarm/wildfly-swarm[WildFly Swarm]
+
+== Requirements
+
+* A deployment should be able to activate a service that requires a capability.
+* The mechanism to let user interact with capabilities should be similar to the one used for services (`ServiceActivator` implementation loaded by Java `ServiceLoader`).
+
+== Non-Requirements
+
+* Current use of `ServiceActivator` is valid and must not be affected by this change.
+* `ServiceActivator` API remains valid and will not be deprecated.
+
+== WildFly Core Requirements
+
+WildFly Core is providing the `CapabilityServiceSupport` to support capability integration in service implementations.
+Capability is a concept from WildFly Core, while `ServiceActivator` is an API provided by MSC. A new API that allows deployments to activate capabilities must be provided by WildFly Core.
+
+* The `CapabilityServiceSupport` API should be exposed to deployments that are activating services so they can integrate with capabilities too.
+* User deployments must be processed and any services that require capabilities be activated.
+
+== Current design
+
+Currently, an user must implement `org.jboss.msc.service.ServiceActivator` and install MSC Services using the `ServiceActivatorContext` parameter.
+Its implementation is loaded by the Java ServiceLoader mechanism and its services are activated during deployment.
+
+[source]
+----
++-----------------------------------+       +--------------------+
+|             << MSC >>             |       |     << User >>     |
+|         ServiceActivator          |<|-----| MyServiceActivator |
++-----------------------------------+       +--------------------+
+| activate(ServiceActivatorContext) |
++-----------------------------------+
+        ^
+       / \
+        |
++---------------------------------------+
+|             << MSC >>                 |
+|       ServiceActivatorContext         |
++---------------------------------------+
+| getServiceTarget(): ServiceTarget     |
+| getServiceRegistry(): ServiceRegistry |
++---------------------------------------+
+----
+
+== Proposed Design
+
+We would need to introduce a new API so that an user could activate services that depends on capabilities.
+This API is provided by WildFly Core (in its wildfly-controller artifact).
+
+The user must implement the `org.jboss.as.controller.CapabilityActivator` interface.
```

Comment:
Should this be in controller? I don't think that module's ever going to not be jboss.api=private as it has too many internals that aren't readily separable.

Other deployment framework classes are in the server module. I think the "deployment framework" classes should be broken out from server, as that module's a mix of internals and subsystem-dev usable stuff, and we should separate that. (I believe that's something fairly straightforward to do.)

This is something I think we need to discuss separately and report back here as it's likely a quite wide ranging conversation.

Reply:
One possibility might be to add this API into MSC with an SPI that WildFly could implement.  Ugly, perhaps, but it'd solve the problem.